### PR TITLE
Refactor containerisation 1

### PR DESCRIPTION
**Pull request by godrei**
### Checklist

- [x] I've read and followed the [Contribution Guidelines](https://github.com/bitrise-io/bitrise/blob/master/.github/CONTRIBUTING.md)
- [x] `README.md` is updated with the changes (if needed)

### Version

Requires a *MAJOR/MINOR/PATCH* [version update](https://semver.org/)

### Context

Containerization-related configuration is getting refactored:

1, Moving service and container configuration to new top-level entities (`services` and `containers`)

From this:

```
workflows:
  primary:
    container:
      image: ruby:3.2
    services:
      postgres:
        image: postgres:13
```

to this:
```
services:
  postgres:
    image: postgres:13

containers:
  ruby:
    image: ruby:3.2

workflows:
  primary:
    container: ruby
    services:
    - postgres
```

2, Step grouping mechanism with a new step list item (`with`)

From this:

```
workflows:
  _setup:
    steps:
    - activate-ssh-key: { }
    - git-clone: {}

  test:
    before_run:
    - _setup
    container: ruby
    services:
    - postgres
    steps:
    - script:
        title: Setup DB
        inputs:
        - content: bundle exec rails db:setup
    - script:
        title: Run tests
        inputs:
        - content: bundle exec rspec
```

to this:

```
workflows:
  test:
    steps:
    - activate-ssh-key@4: { }
    - git-clone@8: {}
    - with:
        container: ruby
        services:
        - postgres
        steps:
        - script:
            title: Setup DB
            inputs:
            - content: bundle exec rails db:setup
        - script:
            title: Run tests
            inputs:
            - content: bundle exec rspec
```

The two changes will be introduced in separate PRs to the `refactor-containerisation` branch.

This PR implements the first change: Moving service and container configuration to new top-level entities (`services` and `containers`)

### Changes

- Introduce new top-level `services` and `containers` entities for configuring service and step execution containers.

### CODE REVIEW

File: models/models_methods.go [hunk 1]
```diff
@@ -268,6 +268,46 @@ func (config *BitriseDataModel) Validate() ([]string, error) {
...
+		}
+		if serviceDef.Image == "" {
+			return nil, fmt.Errorf("service (%s) has no image defined", serviceID)
+		}
+	}
+
+	for workflowID, workflow := range config.Workflows {
+		if workflow.ContainerID != "" {
+			if _, ok := config.Containers[workflow.ContainerID]; !ok {
+				return nil, fmt.Errorf("container (%s) referenced in workflow (%s) but doesn't defined in the config", workflow.ContainerID, workflowID)
```
Comment (ofalvai) [inline models/models_methods.go:293]:
```suggestion
				return nil, fmt.Errorf("container (%s) referenced in workflow (%s), but this container is not defined", workflow.ContainerID, workflowID)
```

File: models/models_methods.go [hunk 2]
```diff
@@ -268,6 +268,46 @@ func (config *BitriseDataModel) Validate() ([]string, error) {
 	}
 	// ---
 
+	// containers
+	for containerID, containerDef := range config.Containers {
+		if containerID == "" {
+			return nil, fmt.Errorf("service (image: %s) has empty ID defined", containerDef.Image)
+		}
+		if containerDef.Image == "" {
```
Comment (ofalvai) [inline models/models_methods.go:276]: nit: I'd also strip whitespace before checking for `""`, it's probably what the Docker API is doing anyway

**Review by ofalvai — APPROVED** [review]

File: models/models_methods.go (hunk 2 above)
Comment (godrei) [reply models/models_methods.go]: Updated in: https://github.com/bitrise-io/bitrise/pull/952/commits/34935bcbef5c72369b8ea1d9fac8a177fc6eca06

File: models/models_methods.go [hunk 3]
```diff
@@ -268,6 +268,46 @@ func (config *BitriseDataModel) Validate() ([]string, error) {
...
+		if workflow.ContainerID != "" {
+			if _, ok := config.Containers[workflow.ContainerID]; !ok {
+				return nil, fmt.Errorf("container (%s) referenced in workflow (%s), but this container is not defined", workflow.ContainerID, workflowID)
+			}
+		}
+
+		serviceIDs := map[string]bool{}
+		for _, serviceID := range workflow.ServiceIDs {
+			if _, ok := config.Services[serviceID]; !ok {
+				return nil, fmt.Errorf("service (%s) referenced in workflow (%s) but doesn't defined in the config", serviceID, workflowID)
```
Comment (ofalvai) [inline models/models_methods.go:300]: One more thing:

```suggestion
				return nil, fmt.Errorf("service (%s) referenced in workflow (%s) but this service is not defined", serviceID, workflowID)
```

Comment (godrei) [reply models/models_methods.go]: Sorry, I forgot to fix this one. Updated in this commit along the related tests: https://github.com/bitrise-io/bitrise/pull/952/commits/28dd70973bd279cd5293b2092520ac832a9d0f00

**Review by ofalvai — APPROVED** [review]

**Pull request merged**